quartical/gains/delay_tec_and_offset/__init__.py

```diff
@@ -0,0 +1,247 @@
+        'MAX_FREQ'
+    )
+)
+
+
+class DelayTecAndOffset(ParameterizedGain):
+
+    solver = staticmethod(delay_tec_and_offset_solver)
+    ms_inputs = ms_inputs
+
+    native_to_converted = (
+        (0, (np.cos,)),
+        (1, (np.sin,)),
+        (1, (no_op,)),
+        (1, (no_op,))
+    )
+    converted_to_native = (
+        (2, trig_to_angle),
+        (1, no_op),
+        (1, no_op)
+    )
+    converted_dtype = np.float64
+    native_dtype = np.float64
+
+    def __init__(self, term_name, term_opts):
+
+        super().__init__(term_name, term_opts)
+
+    @classmethod
+    def _make_freq_map(cls, chan_freqs, chan_widths, freq_interval):
+        # Overload gain mapping construction - we evaluate it in every channel.
+        return np.arange(chan_freqs.size, dtype=np.int32)
+
+    @classmethod
+    def make_param_names(cls, correlations):
+
+        # TODO: This is not dasky, unlike the other functions. Delayed?
+        parameterisable = ["XX", "YY", "RR", "LL"]
+
+        param_corr = [c for c in correlations if c in parameterisable]
+
+        template = ("offset_{}", "tec_{}", "delay_{}")
+
+        return [n.format(c) for c in param_corr for n in template]
+
+    def init_term(self, term_spec, ref_ant, ms_kwargs, term_kwargs, meta=None):
+        """Initialise the gains (and parameters)."""
+
+        gains, gain_flags, params, param_flags = super().init_term(
+            term_spec, ref_ant, ms_kwargs, term_kwargs
+        )
+
+        # Convert the parameters into gains.
+        delay_tec_and_offset_params_to_gains(
+            params,
+            gains,
+            ms_kwargs["CHAN_FREQ"],
+            ms_kwargs["MIN_FREQ"],
+            ms_kwargs["MAX_FREQ"],
+            term_kwargs[f"{self.name}_param_freq_map"],
+        )
+
+        if self.load_from or not self.initial_estimate:
+
+            apply_param_flags_to_params(param_flags, params, 0)
+            apply_gain_flags_to_gains(gain_flags, gains)
+
+            return gains, gain_flags, params, param_flags
+
+        data = ms_kwargs["DATA"]  # (row, chan, corr)
+        flags = ms_kwargs["FLAG"]  # (row, chan)
+        a1 = ms_kwargs["ANTENNA1"]
+        a2 = ms_kwargs["ANTENNA2"]
+        chan_freq = ms_kwargs["CHAN_FREQ"]
+        t_map = term_kwargs[f"{term_spec.name}_time_map"]
+        f_map = term_kwargs[f"{term_spec.name}_param_freq_map"]
+        _, n_chan, n_ant, n_dir, n_corr = gains.shape
+
+        # Rescale the channel frequencies.
+        scale_factor = (chan_freq.min() + chan_freq.max()) / 2
+        scaled_chan_freq = chan_freq / scale_factor
+
+        est_resolution = 0.01  # NOTE: Make this lower.
+
+        # We only need the baselines which include the ref_ant.
+        sel = np.where((a1 == ref_ant) | (a2 == ref_ant))
+        a1 = a1[sel]
+        a2 = a2[sel]
+        t_map = t_map[sel]
+        data = data[sel]
+        flags = flags[sel]
+
+        data[flags == 1] = 0  # Ignore UV-cut, otherwise there may be no est.
```

Comment:
Is this what you meant to do? Doesn't this enforce the UV-cut?

Reply:
QC uses ternary flags (soft, unflagged, hard) corresponding to (-1,0,1). The above only zeros the data at the location of hard flags.